lib/SourceMapDevToolPlugin.js

```diff
@@ -163,233 +161,279 @@ class SourceMapDevToolPlugin {
 					reportProgress(0.0);
 					/** @type {SourceMapTask[]} */
 					const tasks = [];
-					files.forEach((file, idx) => {
-						const asset = compilation.getAsset(file).source;
-						const cache = assetsCache.get(asset);
-						/**
-						 * If presented in cache, reassigns assets. Cache assets already have source maps.
-						 */
-						if (cache && cache.file === file) {
-							for (const cachedFile in cache.assets) {
-								if (cachedFile === file) {
-									compilation.updateAsset(cachedFile, cache.assets[cachedFile]);
-								} else {
-									compilation.emitAsset(cachedFile, cache.assets[cachedFile], {
-										development: true
-									});
-								}
-								/**
-								 * Add file to chunk, if not presented there
-								 */
-								if (cachedFile !== file) {
-									const chunk = fileToChunk.get(file);
-									if (chunk !== undefined) chunk.auxiliaryFiles.add(cachedFile);
-								}
-							}
-							return;
-						}
+					let fileIndex = 0;
+
+					asyncLib.each(
+						files,
+						(file, done) => {
+							const asset = compilation.getAsset(file).source;
+							const cacheIdent = `${compilation.compilerPath}/SourceMapDevToolPlugin/${file}`;
+							const cacheETag = createHash("md4");
+							asset.updateHash(cacheETag);
 
-						reportProgress(
-							(0.5 * idx) / files.length,
-							file,
-							"generate SourceMap"
-						);
-						/** @type {SourceMapTask | undefined} */
-						const task = getTaskForFile(file, asset, options, compilation);
-
-						if (task) {
-							const modules = task.modules;
-
-							for (let idx = 0; idx < modules.length; idx++) {
-								const module = modules[idx];
-								if (!moduleToSourceNameMapping.get(module)) {
-									moduleToSourceNameMapping.set(
-										module,
-										ModuleFilenameHelpers.createFilename(
-											module,
-											{
-												moduleFilenameTemplate: moduleFilenameTemplate,
-												namespace: namespace
-											},
-											{
-												requestShortener,
-												chunkGraph
+							compilation.cache.get(
+								cacheIdent,
+								cacheETag.digest("hex").toString(),
```

Comment:
Looks like changed source map content is not invalidate cache, for example other plugin can potentially change source map

Reply:
Also i think better use https://github.com/webpack/webpack/blob/master/lib/util/createHash.js

/cc @sokra DOes webpack have public API for generate `ETag`? It will be great if webpack respect https://webpack.js.org/configuration/output/#outputhashfunction when you generate hash (i think)

Reply:
> Looks like changed source map content is not invalidate cache, for example other plugin can potentially change source map

Any examples? :)

Reply:
> I think this is a bug in `webpack-sources`. SourceMap content doesn't contribute to hash.

fixed in latest webpack-source version